trap.go

```diff
@@ -207,57 +231,91 @@ func (t *TrapListener) listenUDP(addr string) error {
 			}
 
 			msg := buf[:rlen]
-			traps := t.Params.UnmarshalTrap(msg, false)
-
-			if traps != nil {
-				// Here we assume that t.OnNewTrap will not alter the contents
-				// of the PDU (per documentation, because Go does not have
-				// compile-time const checking).  We don't pass a copy because
-				// the SnmpPacket type is somewhat large, but we could without
-				// violating any implicit or explicit spec.
-				t.OnNewTrap(traps, remote)
-
-				// If it was an Inform request, we need to send a response.
-				if traps.PDUType == InformRequest { //nolint:whitespace
-
-					// Reuse the packet, since we're supposed to send it back
-					// with the exact same variables unless there's an error.
-					// Change the PDUType to the response, though.
-					traps.PDUType = GetResponse
-
-					// If the response can be sent, the error-status is
-					// supposed to be set to noError and the error-index set to
-					// zero.
-					traps.Error = NoError
-					traps.ErrorIndex = 0
-
-					// TODO: Check that the message marshalled is not too large
-					// for the originator to accept and if so, send a tooBig
-					// error PDU per RFC3416 section 4.2.7.  This maximum size,
-					// however, does not have a well-defined mechanism in the
-					// RFC other than using the path MTU (which is difficult to
-					// determine), so it's left to future implementations.
-					ob, err := traps.marshalMsg()
-					if err != nil {
-						return fmt.Errorf("error marshaling INFORM response: %w", err)
-					}
-
-					// Send the return packet back.
-					count, err := t.conn.WriteTo(ob, remote)
-					if err != nil {
-						return fmt.Errorf("error sending INFORM response: %w", err)
-					}
-
-					// This isn't fatal, but should be logged.
-					if count != len(ob) {
-						t.Params.Logger.Printf("Failed to send all bytes of INFORM response!\n")
+			trap, err := t.Params.UnmarshalTrap(msg, false)
+			if err != nil {
+				t.Params.Logger.Printf("TrapListener: error in UnmarshalTrap %s\n", err)
+				continue
+			}
+			if trap.Version == Version3 && trap.SecurityModel == UserSecurityModel && t.Params.SecurityModel == UserSecurityModel {
+				securityParams, ok := t.Params.SecurityParameters.(*UsmSecurityParameters)
+				if !ok {
+					t.Params.Logger.Printf("TrapListener: Invalid SecurityParameters types")
+				}
+				packetSecurityParams, ok := trap.SecurityParameters.(*UsmSecurityParameters)
+				if !ok {
+					t.Params.Logger.Printf("TrapListener: Invalid SecurityParameters types")
+				}
+				snmpEngineID := securityParams.AuthoritativeEngineID
+				msgAuthoritativeEngineID := packetSecurityParams.AuthoritativeEngineID
+				if msgAuthoritativeEngineID != snmpEngineID {
+					if 4 < len(msgAuthoritativeEngineID) && len(msgAuthoritativeEngineID) < 32 {
+						// RFC3414 3.2.3a: continue processing
+					} else {
+						t.usmStatsUnknownEngineIDsCount++
```

Comment:
Is this operation atomic? I guess changing the type to uint32 and casting it to an int in ``reportAuthoritativeEngineID`` enables us to do something like
```go
atomic.AddUint32(&t.usmStatsUnknownEngineIDsCount, 1)
```
which will ensure we are not racing when incrementing this counter on a busy device. What are your thoughts on this?